Added support to handle Temple/Portal in the editor + deharcoded workers.

Bertram25 opened the pull request:
Part of #192

I also added support to handle moving the portal and temple only within the editor.
But this still triggers an assert failure when removing tiles including the one
with the mesh.
E.g.:
ERROR: Assert failed file opendungeons/source/network/ODClient.cpp line 622 info : name=RenderedMovableEntity_Portal2_50

I've made a few tests but so far I really can't tell what's triggering it.
Any advice? ;)

Otherwise, it' working fine.

Commit:
Added support to handle Temple/Portal in the editor.

Part of #192

I also added support to handle moving the portal and temple only within the editor.
But this still triggers an assert failure when removing tiles including the one
with the mesh.
E.g.:
ERROR: Assert failed file opendungeons/source/network/ODClient.cpp line 622 info : name=RenderedMovableEntity_Portal2_50

I've made a few tests but so far I really can't tell what's triggering it.
Any advice? ;)

Otherwise, it' working fine.

hwoarangmy commented:
Concerning the assert/segfault problem, if it only happens when removing tiles, I would guess that you are removing the dungeon temple/portal object twice and that's why it is not found the second time. Could you please post some log ?
If you want, I can have a look tonight.

hwoarangmy commented:
Since the FOW, as you have seen in the PR, the objects creation/remove has become a bit tricky. So, there might be some bug hidding. Moreover, I've not tried to manually remove objects like it is done here. That's why we can also imagine that there is some bug or incompatibility with the architecture.

Bertram25 commented:
> I would guess that you are removing the dungeon temple/portal object twice and that's why it is not found the second time. Could you please post some log ?

Ok, I'll do it, I just wanted to know whether I did something plainly stupid. I do think it is a good way to practice with objects with the latest architecture, though. :)

> That's why we can also imagine that there is some bug or incompatibility with the architecture.

That's the thing I feared. Doing something incompatible with the current architecture. Also, here, I first wanted to only move the temple/portal and I changed the update*Position() accordingly to use a setPosition(there, isMoving = true); on the temple object but this is not working. That's why I changed the logic to delete/recreate the object each time.
The second problem I had is that CreateMeshLocal() is called each time before absorbRoom() if there is an absorption, so I had to let the mesh creation there, while I obtained a duplicate when the room is absorbed.
Maybe you have a better way to do it? (As you basically rewrote that system, eh. ;])

Best regards,

hwoarangmy commented:
Well, Building::updateActiveSpots was supposed to handle active spots. It is overriden in Room and it is the function that removes/creates the activespots when needed (by calling notifyActiveSpotCreated and notifyActiveSpotRemoved). Since the active spots behaviour for Portal/Dungeon temple is pretty specific, IMHO, the easiest would be to override Room::updateActiveSpots (which I forgot to set as virtual btw). IIRC, it is called only after room absorbtion and it works with the editor so it is probably well fitted for that ^^

hwoarangmy commented:
While I'm thinking about it, if you override Room::updateActiveSpots, you should remember that you will have extra calls when a neighboor wall is claimed. That means that you would need to check if the room object has moved before removing/adding it

Bertram25 commented:
Thanks for the hints, I'll redo the patch. :)

Bertram25 commented:
Should be working good now. :)

Commit:
Fixed portal/temple room objects handling thanks to the help of @hwoarangmy :)

- Also made sure to use the central tile to spawn the creature in both rooms.
- Fixed spawning coordinates sometimes wrong in the portal.
- Made sure the Tile::mPosition member is set. (We should decide whether we use it or mX, mY only.)
- Improve a bit the logging of the spawn logic.

Fixes #192

akien-mga commented:
It works fine here. Room merging seems to work great with all kind of room sizes, even though portals and temples are meant to be on 3x3 rooms.

Bertram25 commented:
> It works fine here. Room merging seems to work great with all kind of room sizes, even though portals and temples are meant to be on 3x3 rooms.

Cool :)

> > Ideally, the mode ougt to be passed in as a parameter where it's needed.
> > That's what I've done when I had to but I have to say that updateActivespots is used in every trap/room so that will mean many changes...

So yes, I'd rather add a function in the gamemap class, since it's the true class making the most the separation between the two threads logic. What about:

``` c++
bool GameMap::isInEditorMode() const
{
    if (isServerGameMap())
    {
        return (ODServer::getServerMode == ServerMode::ModeEditor)
    }

    return (ODFrameListener::getSingleton().getModeManager->getCurrentModeTypeExceptConsole() == Modetype::EDITOR)
}
```

Commit:
Dehardcoded the default worker spawning logic.

- Made the workers part of the spawn pools to permit using some generic logic to spawn them
  in the temples.
- I also used @oyvindln function to test for the editor for now until it is better fixed.
- Now the hero faction will spawn dwarf workers instead of kobold, and this is configurable.

Commit:
Small API update.

Made a function parameter more understandable.
As discussed with @hwoarangmy some time ago.

Commit:
Slightly decrease the in-game zoom in/out speed.

This was a bit too sensitive, especially using the mouse.
Fortunately, once the option menu is there, this should become configurable. :)

hwoarangmy commented:
> I'd rather add a function in the gamemap class, since it's the true class making the most the separation between the two threads logic

That seems ok to me

Bertram25 commented:
Note to self:
- Add function to test whether one is in map editing mode in the game map object.
- Add support to define a default worker class in Seat, when initialization the spawn pool.

I'll do that asap. :)

Commit:
Made the faction default worker spawned be defined in a specific parameter in the factions.cfg file.

This is much cleaner, indeed. Thanks to @hwoarangmy for pointing out.

Commit:
Introduced the GameMap::isInEditorMode() function.

Used to know when we are editing a map, while doing all the dirty logic.

Thanks to @hwoaragmy for the original idea.

Commit:
Misc fixes requested by @hwoarangmy

hwoarangmy commented:
If I am not mistaken, you have done all you wanted in this PR, right ?
If it is the case, I will test tonight ^^

Commit:
Doc fix, as pointed out by @hwoarangmy

Bertram25 commented:
Yep, normally I'm done. :)

hwoarangmy commented:
Seems like rebasing is needed ^^

Bertram25 commented:
k, I do that asap. :)

Bertram25 commented:
Rebased :)

hwoarangmy commented:
Nice. I will test tonight

oyvindln commented on source/utils/ConfigManager.h:
```diff
@@ -130,6 +134,10 @@ class ConfigManager : public Ogre::Singleton<ConfigManager>
+
+    //! \brief Stores the faction default worker creature definition.
+    std::map<const std::string, std::string> mFactionDefaultWorkerClass;
+
```
Not that it's going to make a huge difference, but the strings in the map probably shouldn't be const, as that may prevent move-construction if the map has to reallocate memory.

hwoarangmy replied:
For maps? I thought it was for vectors only...

oyvindln replied:
Hm, you are right about maps it seems.

hwoarangmy commented:
Works well, I merge
